timeprovider.go

```diff
@@ -4,11 +4,27 @@ import (
 	"time"
 )
 
-// TimeProvider defines the genesis time of slot 0 and allows to convert index to and from time.
+// TimeProvider defines the perception of time, slots and epochs.
+// It allows to convert slots to and from time, and epochs to and from slots.
+// Slots are counted starting from 1 because 0 is reserved for the genesis which has to be addressable as its own slot.
+// Epochs are counted starting from 0.
+//
+// Example: with slotDurationSeconds = 10 and slotsPerEpochExponent = 3
+// <> inclusive range boundary, () exclusive range boundary
+// slot 0: <-inf; genesis)
+// slot 1: <genesis; genesis+10)
+// slot 2: <genesis+10; genesis+20)
+// ...
+// epoch 0: <slot 0; slot 8)
+// epoch 1: <slot 8; slot 16)
+// epoch 2: <slot 16; slot 24)
+// ...
```

Comment:
```suggestion
// Example: with slotDurationSeconds = 10 and slotsPerEpochExponent = 3
// [] inclusive range boundary, () exclusive range boundary
// slot 0: [-inf; genesis)
// slot 1: [genesis; genesis+10)
// slot 2: [genesis+10; genesis+20)
// ...
// epoch 0: [slot 0; slot 8)
// epoch 1: [slot 8; slot 16)
// epoch 2: [slot 16; slot 24)
// ...
```

Nit: The conventional way to write inclusive and exclusive ranges is with `[]` and `()` respectively. Would be a bit nicer to use that here.